--- XML/Filter/GenericChunk.pm.orig	2013-04-12 23:04:56.949954642 +0200
+++ XML/Filter/GenericChunk.pm	2013-04-13 01:33:02.000000000 +0200
@@ -39,34 +39,39 @@
 }
 
 sub start_element {
-    my $self = shift;
-    my $element = shift;
-    $self->_init_element($element);
+    my ($self,$element) = @_;
+
+    unless( $self->is_tag() ) {
+        $self->_init_element($element);
+    }
 
-    unless ( $self->is_tag() and $self->{DropElement} == 1 ) { 
+    # the second is_tag() is needed.
+    # guten tag?
+    unless ( $self->{DropElement} == 1 and $self->is_tag() ) {
         $self->SUPER::start_element($element);
     }
+
 }
 
 sub end_element {
-    my $self = shift;
-    my $element = shift;
+    my ($self,$element) = @_;
+    #my $element = shift;
 
     # need to remember if this is the active tag, because after reset this 
     # information is not available anymore
     my $istag = $self->is_tag();
 
-    $self->_reset_element($element);
+    $self->_reset_element($element) if $istag;
     $self->reset_data();
 
-    unless ( $istag and $self->{DropElement} == 1 ) { 
+    unless ( $istag and $self->{DropElement} == 1 ) {
         $self->SUPER::end_element($element);
     }
 }
 
 sub relaxed_names {
     my $self = shift;
-    if ( scalar @_ && defined $_[0] ) {
+    if ( defined $_[0] ) {
         $self->{RelaxedNames} = shift;
     }
     return $self->{RelaxedNames};
@@ -79,23 +84,24 @@
 }
 
 sub set_namespace {
-    my $self = shift;
-    my $uri = shift;
+    my ($self,$uri) = @_;
+    #my $uri = shift;
     $self->{NamespaceURI} = $uri if defined $uri;
 }
 
 sub reset_tagname {
-    my $self = shift;
+    my ($self) = $_;
     $self->{TagName} = [];
     $self->{TagByName} = {};
 }
 
 sub is_tag {
-    return length $_[0]->{CurrentElement} > 0 ? 1 : 0;
+    #return length $_[0]->{CurrentElement} > 0 ? 1 : 0;
+    $_[0]->{CurrentElement} ? 1 : 0;
 }
 
 sub flush_chunk {
-    my $self = shift;
+    my ($self) = @_;
 
     my $docfrag = $self->get_data_fragment();
     if ( defined $docfrag and defined $docfrag->childNodes() ) {
@@ -145,16 +151,18 @@
 sub reset_data { $_[0]->{WBChunk} = ""; }
 
 sub _prepare_names {
-    my $self = shift;
+    my ($self) = @_;
     # this precaches the tagnames
     map {$self->{TagByName}->{$_} = 1;} @{$self->{TagName}};
 }
 
 sub _init_element {
-    my $self = shift;
-    my $element = shift;
+    my ($self, $element) = @_;
+
+    # my $self = shift;
+    # my $element = shift;
 
-    unless( length $self->{CurrentElement} > 0 ) {
+    #unless( $self->{CurrentElement} ) {
         # in this case we test the entire name!
         my $name = "";
         if ( $self->relaxed_names() == 1 ) {
@@ -175,14 +183,14 @@
         if ( defined $name and exists $self->{TagByName}->{$name} ) {
             $self->{CurrentElement} = $name;
         }
-    }
+    #}
 }
 
 sub _reset_element {
-    my $self = shift;
-    my $element = shift;
+    my ($self,$element) = @_;
+    #my $element = shift;
 
-    if ( $self->is_tag() ) {
+    #if ( $self->is_tag() ) {
         my $name = "";
         if ( $self->relaxed_names() == 1) {
             $name = $element->{Name};
@@ -205,7 +213,7 @@
              and $self->{CurrentElement} eq $name ) {
             $self->{CurrentElement} = "";
         }
-    }
+    #}
 }
 
 1;
